File: packages/types/src/integration.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

// All the app integrations that are available
interface IAppIntegrationMetadata {
  auth_provider?: string;
  community?: boolean;
  install_handler?: string;
  install_url?: string | null;
  installed_description?: string;
  not_installed_description?: string;
  project_description?: string;
  project_logo_url?: string | null;
  supports_project_sync?: boolean;
  [key: string]: unknown;
}

interface IAppIntegration {
  author: string;
  avatar_url: string | null;
  created_at: string;
  created_by: string | null;
  description: any;
  id: string;
  metadata: IAppIntegrationMetadata;
  network: number;
  provider: string;
  redirect_url: string;
  title: string;
  updated_at: string;
  updated_by: string | null;
  verified: boolean;
  webhook_secret: string;
  webhook_url: string;
}

interface IWorkspaceIntegration {
  actor: string;
  api_token: string;
  config: any;
  created_at: string;
  created_by: string;
  id: string;
  integration: string;
  integration_detail: IAppIntegration;
  metadata: any;
  updated_at: string;
  updated_by: string;
  workspace: string;
}

// slack integration
interface ISlackIntegration {
  id: string;
  created_at: string;
  updated_at: string;
  access_token: string;
  scopes: string;
  bot_user_id: string;
  webhook_url: string;
  data: ISlackIntegrationData;
  team_id: string;
  team_name: string;
  created_by: string;
  updated_by: string;
  project: string;
  workspace: string;
  workspace_integration: string;
}

interface ISlackIntegrationData {
  ok: boolean;
  team: {
    id: string;
    name: string;
  };
  scope: string;
  app_id: string;
  token_type: string;
  authed_user: string;
  bot_user_id: string;
  access_token: string;
  incoming_webhook: {
    url: string;
    channel: string;
    channel_id: string;
    configuration_url: string;
  };
}

export type {
  IAppIntegration,
  IAppIntegrationMetadata,
  ISlackIntegration,
  ISlackIntegrationData,
  IWorkspaceIntegration,
};

File: apps/api/plane/app/serializers/integration.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Third party imports
from rest_framework import serializers

# Module imports
from .base import DynamicBaseSerializer
from plane.db.models import GithubRepository, GithubRepositorySync, Integration, WorkspaceIntegration


class IntegrationSerializer(DynamicBaseSerializer):
    metadata = serializers.SerializerMethodField()

    class Meta:
        model = Integration
        fields = "__all__"
        read_only_fields = [
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]

    def get_metadata(self, obj):
        metadata = dict(obj.metadata or {})

        if obj.redirect_url and "install_url" not in metadata:
            metadata["install_url"] = obj.redirect_url

        return metadata


class WorkspaceIntegrationSerializer(DynamicBaseSerializer):
    integration_detail = IntegrationSerializer(source="integration", read_only=True)

    class Meta:
        model = WorkspaceIntegration
        fields = "__all__"
        read_only_fields = [
            "workspace",
            "actor",
            "integration",
            "api_token",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]


class GithubRepositorySerializer(DynamicBaseSerializer):
    full_name = serializers.SerializerMethodField()
    html_url = serializers.CharField(source="url", read_only=True)

    class Meta:
        model = GithubRepository
        fields = [
            "id",
            "repository_id",
            "name",
            "owner",
            "url",
            "html_url",
            "full_name",
            "config",
            "project",
            "workspace",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "project",
            "workspace",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]

    def get_full_name(self, obj):
        return f"{obj.owner}/{obj.name}"


class GithubRepositorySyncSerializer(DynamicBaseSerializer):
    repo_detail = GithubRepositorySerializer(source="repository", read_only=True)

    class Meta:
        model = GithubRepositorySync
        fields = [
            "id",
            "repository",
            "repo_detail",
            "credentials",
            "actor",
            "workspace_integration",
            "label",
            "project",
            "workspace",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "repository",
            "repo_detail",
            "credentials",
            "actor",
            "workspace_integration",
            "label",
            "project",
            "workspace",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]
